api/src/main/java/uk/gov/hmcts/payment/api/controllers/PaymentStatusController.java

```diff
@@ -0,0 +1,159 @@
+        @ApiResponse(code = 429, message = "Request already received for this failure reference"),
+        @ApiResponse(code = 500, message = "Internal Server Error")
+
+    })
+    @PaymentExternalAPI
+    @PostMapping(path = "/payment-failures/bounced-cheque")
+    public ResponseEntity<String> paymentStatusBouncedCheque(@Valid @RequestBody PaymentStatusBouncedChequeDto paymentStatusBouncedChequeDto){
+
+        boolean psuLockFeature = featureToggler.getBooleanValue("payment-status-update-flag",false);
+        LOG.info("feature toggler enable for  bounced-cheque : {}",psuLockFeature);
+
+        if(psuLockFeature){
+            return new ResponseEntity<>("service unavailable", HttpStatus.SERVICE_UNAVAILABLE);
+        }
+
+        LOG.info("Received payment status request bounced-cheque : {}", paymentStatusBouncedChequeDto);
+        Optional<Payment> payment = paymentRepository.findByReference(paymentStatusBouncedChequeDto.getPaymentReference());
+
+        if(payment.isEmpty()){
+            throw new PaymentNotFoundException("No Payments available for the given Payment reference");
+        }
+
+        Optional<PaymentFailures> paymentFailures = paymentStatusUpdateService.searchFailureReference(paymentStatusBouncedChequeDto.getFailureReference());
+
+        if(paymentFailures.isPresent()){
+            throw new FailureReferenceNotFoundException("Request already received for this failure reference");
+        }
+
+         PaymentFailures insertPaymentFailures =  paymentStatusUpdateService.insertBounceChequePaymentFailure(paymentStatusBouncedChequeDto);
+
+          if(null != insertPaymentFailures.getId()){
+              paymentStatusUpdateService.cancelFailurePaymentRefund(paymentStatusBouncedChequeDto.getPaymentReference());
+            }
+
+        return new ResponseEntity<>("successful operation", HttpStatus.OK);
+
+    }
+
+    @PaymentExternalAPI
+    @PostMapping(path = "/payment-failures/chargeback")
+    public ResponseEntity<String> paymentStatusChargeBack(@Valid @RequestBody PaymentStatusChargebackDto paymentStatusChargebackDto){
+
+        boolean psuLockFeature = featureToggler.getBooleanValue("payment-status-update-flag",false);
+        LOG.info("feature toggler enable for  chargeback : {}",psuLockFeature);
+
+        if(psuLockFeature){
+            return new ResponseEntity<>("service unavailable", HttpStatus.SERVICE_UNAVAILABLE);
+        }
+
+        LOG.info("Received payment status request chargeback : {}", paymentStatusChargebackDto);
+        Optional<Payment> payment = paymentRepository.findByReference(paymentStatusChargebackDto.getPaymentReference());
+
+        if(payment.isEmpty()){
+            throw new PaymentNotFoundException("No Payments available for the given Payment reference");
+        }
+
+        Optional<PaymentFailures> paymentFailures = paymentStatusUpdateService.searchFailureReference(paymentStatusChargebackDto.getFailureReference());
```

Comment:
Why are 2 calls being made to Payment Failures table - one for checking the failure reference record and then inserting if not found? why cannot it be done in 1 call, if record already exists then throw an exception.

Reply:
implemented change